core/src/main/scala/cats/free/Codensity.scala

```diff
@@ -0,0 +1,64 @@
+package cats
+package free
+
+/** 
+ * Codensity Monad
+ * 
+ * Codensity is used to improve quadratic complexity of free monadic structures.
+ * This approach is presented in this paper:
+ * http://www.janis-voigtlaender.eu/papers/AsymptoticImprovementOfComputationsOverFreeMonads.pdf
+ *
+ * The idea behind Codensity is simply to replace left associated structures by wrapping into right associated codensity.
+ * There is an isomorphism between Free monadic structure and codensity monad making it lossless to be used in your code.
+ *
+ * Interesting aspects to know:  
+ * - Codensity[F[_], A] = NaturalTransformation[λ[α => Kleisli[F, A, α]], F]
+ * - For any type constructor F[_], Codensity is a monad
+ */
+trait Codensity[F[_], A] {
```

Comment:
I think the main argument for an abstract class is binary compatibility. I believe you can safely add a concrete method to an abstract class but not to a trait. See [this SO post](http://stackoverflow.com/questions/18366817/is-adding-a-trait-method-with-implementation-breaking-backward-compatibility). I believe @mpilquist said this wasn't the case a little while back, so he might know something I don't.

I believe there could also be a slight performance gain since it avoids delegation to another method, but I'm not very good at reasoning about JVM performance.